app/plugin.go

```diff
@@ -407,22 +416,93 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
 	return resp, nil
 }
 
+func (a *App) InstallMarketplacePlugin(request *model.InstallMarketplacePluginRequest) (*model.Manifest, *model.AppError) {
+	var pluginFile, signatureFile io.ReadSeeker
+
+	prepackagedPlugin, appErr := a.getPrepackagedPlugin(request.Id, request.Version)
+	if appErr != nil && appErr.Id != "app.plugin.marketplace_plugins.not_found.app_error" {
+		return nil, appErr
+	} else if prepackagedPlugin != nil {
```

Comment:
Nit: Do we need to return an error if there is no plugin found in the prepackaged? Not sure what is the value of returning a `app.plugin.marketplace_plugins.not_found.app_error` which we will not use and still have to check if `prepackagedPlugin != nil`

Either way I think we can drop this `else` here

Reply:
Had a lengthy discussion with @lieut-data and we settled on to follow this paradigm where we `return nil, errors.New("entity not found")` when fetching a single entity to notify the caller that it failed to find the specific entity. And to `return nil, nil` when fetching a slice of items. This pattern is commonly seen in the Go SQL package.

[QueryRow](https://golang.org/pkg/database/sql/#Stmt.QueryRow) returns `ErrNoRows` if no row is found.
[Query](https://golang.org/pkg/database/sql/#DB.Query) returns `*Rows` that has the semantics of `return nil, nil` when returning multiple rows.

JS's [Lodash](https://lodash.com/docs/#findIndex) has similar semantics for `_.findIndex([...], ...)` vs `_.filter([...], ...)` as well.

Because of that it's possible that `prepackagedPlugin` is `nil` when `appErr.Id == "app.plugin.marketplace_plugins.not_found.app_error"` in which case I can't use `prepackagedPlugin`, and therefore this check.

Reply:
Understood why the AppError. 

Nit: I meant drop the `else` but keep the condition check given that the previous condition returns. 
 
```suggestion
	if prepackagedPlugin != nil {
```

Reply:
ahh true, thought my `prepackagedPlugin` was inside the if, good call. Updated.